homeassistant/components/canary/config_flow.py

```diff
@@ -0,0 +1,117 @@
+"""Config flow for Canary."""
+import logging
+from typing import Any, Dict, Optional
+
+from canary.api import Api
+from requests import ConnectTimeout, HTTPError
+import voluptuous as vol
+
+from homeassistant.config_entries import CONN_CLASS_CLOUD_POLL, ConfigFlow, OptionsFlow
+from homeassistant.const import CONF_PASSWORD, CONF_TIMEOUT, CONF_USERNAME
+from homeassistant.core import callback
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+
+from .const import CONF_FFMPEG_ARGUMENTS, DEFAULT_FFMPEG_ARGUMENTS, DEFAULT_TIMEOUT
+from .const import DOMAIN  # pylint: disable=unused-import
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def validate_input(hass: HomeAssistantType, data: dict) -> Dict[str, Any]:
+    """Validate the user input allows us to connect.
+
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+    # constructor does login call
+    Api(
+        data[CONF_USERNAME],
+        data[CONF_PASSWORD],
+        data[CONF_TIMEOUT],
+    )
+
+    return True
+
+
+class CanaryConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Canary."""
+
+    VERSION = 1
+    CONNECTION_CLASS = CONN_CLASS_CLOUD_POLL
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Get the options flow for this handler."""
+        return CanaryOptionsFlowHandler(config_entry)
+
+    async def async_step_import(
+        self, user_input: Optional[ConfigType] = None
+    ) -> Dict[str, Any]:
+        """Handle a flow initiated by configuration file."""
+        return await self.async_step_user(user_input)
+
+    async def async_step_user(
+        self, user_input: Optional[ConfigType] = None
+    ) -> Dict[str, Any]:
+        """Handle a flow initiated by the user."""
+        if self._async_current_entries():
+            return self.async_abort(reason="single_instance_allowed")
+
+        errors = {}
+
+        if user_input is not None:
+            if CONF_TIMEOUT not in user_input:
+                user_input[CONF_TIMEOUT] = DEFAULT_TIMEOUT
+
+            try:
+                await self.hass.async_add_executor_job(
+                    validate_input, self.hass, user_input
+                )
+            except (ConnectTimeout, HTTPError):
+                errors["base"] = "cannot_connect"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                return self.async_abort(reason="unknown")
+            else:
+                return self.async_create_entry(
+                    title=user_input[CONF_USERNAME],
+                    data=user_input,
+                )
+
+        data_schema = {
+            vol.Required(CONF_USERNAME): str,
+            vol.Required(CONF_PASSWORD): str,
+        }
+
+        if self.show_advanced_options:
+            data_schema[vol.Optional(CONF_TIMEOUT, default=DEFAULT_TIMEOUT)] = int
```

Comment:
The config form should only contain required attributes and use reasonable defaults. Timeout would be better suited as an option.
What's more interesting is this is never presented to the user. So it probably should not be here. If it's needed for testing, there are better ways to set the timeout.

Reply:
this would be presented only if you are in HAs advanced mode. not shown to common user ie falling back to default. i guess we dont need it for config entry but the yaml did offer such control so thats why it exists